File: src/agents/collector.py
import asyncio
import os
from datetime import datetime, timezone
from src.sources.who import WHOSource
from src.sources.cdc import CDCSource
from src.sources.nih import NIHSource
from src.sources.pubmed import PubMedSource
from src.sources.medicalxpress import MedicalXpressSource
from src.sources.google_news import GoogleNewsSource
from src.sources.reuters import ReutersSource
from src.sources.kdca import KDCASource
from src.agents.crawler import crawl_article
from src.agents.summarizer import summarize
from src.agents.classifier import classify
from src.utils.text_cleaner import compute_hash
from src.db.client import get_client
from src.db.models import NewsArticle
from src.utils.logger import logger

MAX_PER_SOURCE = int(os.getenv("MAX_ARTICLES_PER_SOURCE", "50"))

ALL_SOURCES = [
    WHOSource(), CDCSource(), NIHSource(), PubMedSource(),
    MedicalXpressSource(), GoogleNewsSource(), ReutersSource(), KDCASource(),
]


async def _fetch_all_sources() -> list:
    tasks = [src.fetch() for src in ALL_SOURCES]
    results = await asyncio.gather(*tasks, return_exceptions=True)
    articles = []
    for i, result in enumerate(results):
        src_name = ALL_SOURCES[i].name
        if isinstance(result, Exception):
            logger.error(f"[Collector] {src_name} 수집 실패: {result}")
        else:
            sliced = result[:MAX_PER_SOURCE]
            logger.info(f"[Collector] {src_name}: {len(sliced)}건 수집")
            articles.extend(sliced)
    return articles


def _deduplicate(articles: list) -> list:
    seen_hashes = set()
    seen_urls = set()
    unique = []
    for a in articles:
        h = compute_hash(a.title, a.summary or "")
        if h in seen_hashes or a.url in seen_urls:
            continue
        seen_hashes.add(h)
        seen_urls.add(a.url)
        a._hash = h
        unique.append(a)
    logger.info(f"[Collector] 중복 제거 후: {len(unique)}건 (원본: {len(articles)}건)")
    return unique


async def _is_already_stored(url: str) -> bool:
    try:
        db = get_client()
        res = db.table("news_articles").select("id").eq("url", url).limit(1).execute()
        return len(res.data) > 0
    except Exception:
        return False


async def _process_article(raw_article) -> dict | None:
    """단일 기사 크롤링 → 요약 → 분류 → DB 저장."""
    if not raw_article.title or not raw_article.url:
        return None

    if await _is_already_stored(raw_article.url):
        logger.debug(f"[Collector] 이미 저장됨: {raw_article.url}")
        return None

    content = await crawl_article(raw_article.url)
    text_for_llm = content or raw_article.summary or ""

    llm_result = await summarize(raw_article.title, text_for_llm, raw_article.language)
    cls_result = await classify(
        raw_article.title,
        llm_result.get("summary", ""),
        llm_result.get("diseases", []),
    )

    article = NewsArticle(
        source=raw_article.source,
        title=raw_article.title,
        url=raw_article.url,
        published_at=raw_article.published_at,
        content=content,
        summary=llm_result.get("summary", ""),
        keywords=llm_result.get("keywords", []),
        diseases=llm_result.get("diseases", []),
        risk_level=cls_result.get("risk_level", "unknown"),
        category=cls_result.get("category", "other"),
        language=raw_article.language,
        is_processed=True,
        content_hash=compute_hash(raw_article.title, content),
    )

    try:
        db = get_client()
        db.table("news_articles").upsert(
            article.model_dump(mode="json"),
            on_conflict="url",
        ).execute()
        logger.info(f"[Collector] 저장: [{cls_result.get('risk_level','?').upper()}] {raw_article.title[:60]}")
        return article.model_dump()
    except Exception as e:
        logger.error(f"[Collector] DB 저장 오류: {e}")
        return None


async def run_collection() -> dict:
    """전체 수집 파이프라인 실행."""
    start = datetime.now(timezone.utc)
    logger.info("=== 의료 뉴스 수집 시작 ===")

    raw_articles = await _fetch_all_sources()
    unique_articles = _deduplicate(raw_articles)

    # 동시 처리 (세마포어로 LLM 과부하 방지)
    sem = asyncio.Semaphore(5)

    async def bounded(article):
        async with sem:
            return await _process_article(article)

    results = await asyncio.gather(*[bounded(a) for a in unique_articles], return_exceptions=True)
    saved = [r for r in results if r and not isinstance(r, Exception)]

    elapsed = (datetime.now(timezone.utc) - start).total_seconds()
    logger.info(f"=== 수집 완료: {len(saved)}건 저장, {elapsed:.1f}초 ===")

    return {
        "total_fetched": len(raw_articles),
        "unique": len(unique_articles),
        "saved": len(saved),
        "elapsed_seconds": elapsed,
    }

File: src/agents/collector.ts
import { ALL_SOURCES } from '@/sources';
import type { RawArticle } from '@/sources';
import { crawlArticle } from './crawler';
import { analyzeArticle } from './analyzer';
import { computeHash } from '@/lib/hash';
import { getSupabaseClient } from '@/lib/supabase';
import { logger } from '@/lib/logger';
import type { NewsArticleInsert, CollectionResult } from '@/types';

const MAX_PER_SOURCE = parseInt(process.env.MAX_ARTICLES_PER_SOURCE ?? '50');
// 무료 LLM 분당 16회 제한 → 동시 3개 + 배치 간 12초 대기 = 분당 ~15건
const CONCURRENCY = 3;
const BATCH_DELAY_MS = 12000;

// ── Source fetching ───────────────────────────────────────────────────────────

async function fetchAllSources(): Promise<RawArticle[]> {
  const results = await Promise.allSettled(ALL_SOURCES.map((s) => s.fetch()));
  const articles: RawArticle[] = [];

  for (let i = 0; i < results.length; i++) {
    const r = results[i];
    const name = ALL_SOURCES[i].name;
    if (r.status === 'fulfilled') {
      const sliced = r.value.slice(0, MAX_PER_SOURCE);
      logger.info(`[Collector] ${name}: ${sliced.length}건`);
      articles.push(...sliced);
    } else {
      logger.error(`[Collector] ${name} 실패: ${r.reason}`);
    }
  }
  return articles;
}

// ── Deduplication ─────────────────────────────────────────────────────────────

function deduplicate(articles: RawArticle[]): Array<RawArticle & { _hash: string }> {
  const seenHash = new Set<string>();
  const seenUrl = new Set<string>();
  const unique: Array<RawArticle & { _hash: string }> = [];

  for (const a of articles) {
    if (!a.title || !a.url) continue;
    const hash = computeHash(a.title, a.summary ?? '');
    if (seenHash.has(hash) || seenUrl.has(a.url)) continue;
    seenHash.add(hash);
    seenUrl.add(a.url);
    unique.push({ ...a, _hash: hash });
  }

  logger.info(`[Collector] 중복 제거: ${articles.length} → ${unique.length}건`);
  return unique;
}

// ── DB duplicate check (요약 없는 기사는 재처리 허용) ─────────────────────────

async function isStored(url: string): Promise<boolean> {
  try {
    const db = getSupabaseClient();
    const { data } = await db
      .from('news_articles')
      .select('id, summary')
      .eq('url', url)
      .limit(1);
    if (!data || data.length === 0) return false;
    // 요약이 있는 기사만 "이미 처리됨"으로 간주
    return typeof (data[0] as Record<string, unknown>).summary === 'string' &&
      ((data[0] as Record<string, unknown>).summary as string).length > 10;
  } catch {
    return false;
  }
}

// ── Single article pipeline ───────────────────────────────────────────────────

async function processArticle(
  raw: RawArticle & { _hash: string }
): Promise<boolean> {
  if (await isStored(raw.url)) {
    logger.debug(`[Collector] 이미 저장됨: ${raw.url}`);
    return false;
  }

  const content = await crawlArticle(raw.url);
  const textForLLM = content || raw.summary || raw.title;

  // 요약 + 분류를 단일 LLM 호출로 처리
  const analysis = await analyzeArticle(textForLLM ? `${raw.title}\n\n${textForLLM}` : raw.title, '', raw.language);

  const record: NewsArticleInsert = {
    source: raw.source,
    title: raw.title,
    url: raw.url,
    published_at: raw.published_at?.toISOString() ?? null,
    content: content || null,
    summary: analysis.summary || null,
    keywords: analysis.keywords,
    diseases: analysis.diseases,
    risk_level: analysis.risk_level,
    category: analysis.category,
    language: raw.language,
    is_processed: true,
    content_hash: computeHash(raw.title, content),
  };

  try {
    const db = getSupabaseClient();
    const { error } = await db
      .from('news_articles')
      .upsert(record, { onConflict: 'url' });

    if (error) throw error;

    const riskEmoji = { critical: '🔴', high: '🟠', medium: '🟡', low: '🟢', unknown: '⚪' }[analysis.risk_level];
    logger.info(`[Collector] ${riskEmoji} [${analysis.risk_level}][${analysis.category}] ${raw.title.slice(0, 55)}`);
    return true;
  } catch (e) {
    logger.error(`[Collector] DB 저장 오류: ${e}`);
    return false;
  }
}

// ── Concurrency helper (배치 간 지연 포함) ─────────────────────────────────────

async function runConcurrent<T, R>(
  items: T[],
  concurrency: number,
  fn: (item: T) => Promise<R>
): Promise<R[]> {
  const results: R[] = [];
  for (let i = 0; i < items.length; i += concurrency) {
    const batch = items.slice(i, i + concurrency);
    const batchResults = await Promise.allSettled(batch.map(fn));
    for (const r of batchResults) {
      if (r.status === 'fulfilled') results.push(r.value);
    }
    // 마지막 배치가 아니면 대기 (rate limit 준수)
    if (i + concurrency < items.length) {
      await new Promise((r) => setTimeout(r, BATCH_DELAY_MS));
    }
  }
  return results;
}

// ── Main entry ────────────────────────────────────────────────────────────────

export async function runCollection(): Promise<CollectionResult> {
  const start = Date.now();
  logger.info('=== 의료 뉴스 수집 시작 ===');

  const rawArticles = await fetchAllSources();
  const unique = deduplicate(rawArticles);

  const saved = await runConcurrent(unique, CONCURRENCY, processArticle);
  const savedCount = saved.filter(Boolean).length;

  const elapsed = (Date.now() - start) / 1000;
  logger.info(`=== 수집 완료: ${savedCount}건 저장 / ${elapsed.toFixed(1)}초 ===`);

  return {
    total_fetched: rawArticles.length,
    unique: unique.length,
    saved: savedCount,
    elapsed_seconds: elapsed,
  };
}
